### PR TITLE
Lamps 2

### PR DESCRIPTION
**Pull request by Tictim**
## What
This PR adds lamps.

This PR is continuation of #1216. It adds lamps, more specifically, multiple variations of colored lamps.

## Implementation Details
In addition to the 16 variations of lamps for each dye colors, this PR adds 4 category of variations for lamps.

* The leftmost block is a normal variation. Normal lamps are on by default. It gets turned off when powered with redstone.
* Second block from the left is an *inverted* variation. Instead of being turned on by default, inverted lamps are turned off by default. Providing redstone signal turns the lamp on.
* The middle one is a *borderless* variation. It is a lamp block with plain color texture.
* Second to last one is *no bloom* variation. No bloom lamps don't provide any bloom effect. The lamp is still rendered with full brightness.
* The last one is *no light* variation. No light lamps don't provide functional lightings, instead they only provide the visual effect.

Each 4 of the variation can be applied simultaneously, like "borderless inverted" or "no bloom, no light". As a result this PR creates 16 possible variations of lamp for one color, or 128 total lamp variations. EDIT: 256 total variations. I can't do math.

To ease the distinction between visually identical variations of lamps, a small item overlay was implemented. This design is not final, and I would appreciate any feedback regarding this feature.

Along with numerous implementation changes, this PR also changes texture of the lamps. Color of individual lamps were also adjusted to provide better visuals.

## Outcome
Adds lamps.

## Additional Information
I don't have any idea how this would behave with Optifine. At least it should work without CTM.

As this PR changes some stuff on ASM, compatibility between endercore should be re-checked.

## Potential Compatibility Issues
None

### CODE REVIEW

**Comment by LAGIdiot** [comment]
This is really interesting addition. And even though I am not into decoration stuff I like it!

I see quite some potential for additional stuff in future like intelligent lamps, controller for intelligent lamps (with programmability?), app for intelligent lamps and so on.

If I am not wrong all this should be available from LV (Assembler, chemical dyes), right?

Maybe it should be moved to new "decoration" module. That should be possible, right @DStrand1 ?

Potential problem I am seeing with this is JEI pollution as it is 256 items.

**Review by ALongStringOfNumbers — COMMENTED** [review]
Is it worth having entirely new blocks just for the inverted state, or could the inverted state be something that you just place any light into a crafting grid and you get back the same light state, but inverted?

I am asking because just having the inverted state be a separate block for each light doubles the amount of blocks that are added, when having the inverted state as a property that you get from crafting the light could reduce this amount.

**Comment by Tictim** [comment]
To make lamps toggle-able with redstone the extra set of blocks are required. All alternative approaches I am aware of (tile entities + actual states, extended states) are suboptimal.

At least "powered" blocks are not registered as items.

File: src/main/java/gregtech/common/blocks/BlockLamp.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+public class BlockLamp extends VariantBlock<EnumDyeColor> {
```
[the remark this answers is not in the document]
Comment (Tictim) [reply src/main/java/gregtech/common/blocks/BlockLamp.java:44]: Did I missed something?

File: src/main/java/gregtech/client/model/BorderlessLampBakedModel.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public BorderlessLampBakedModel(ModelResourceLocation modelLocation) {
+        super(modelLocation);
+    }
+
+    public BorderlessLampBakedModel(IBakedModel model) {
+        super(model);
+    }
+
+    @Override
+    public List<BakedQuad> getQuads(@Nullable IBlockState state, @Nullable EnumFacing side, long rand) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/client/model/BorderlessLampBakedModel.java:29]: Why doesn't this have to go through the same filtering system as the non borderless model?

File: src/main/java/gregtech/common/blocks/BlockLamp.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    @Nonnull
+    @Override
+    protected BlockStateContainer createBlockState() {
+        this.VARIANT = BlockColored.COLOR;
+        this.VALUES = EnumDyeColor.values();
+        return new BlockStateContainer(this, VARIANT);
+    }
+
+    @Nonnull
+    protected IBlockState getComplementaryState(IBlockState state) {
```
[the remark this answers is not in the document]
Comment (Tictim) [reply src/main/java/gregtech/common/blocks/BlockLamp.java]: Well it's removed entirely 😃

File: src/main/java/gregtech/client/model/BorderlessLampBakedModel.java (hunk 2 above)
Comment (Tictim) [reply src/main/java/gregtech/client/model/BorderlessLampBakedModel.java:30]: Bordered lamps have two layers to draw (SOLID for border/base, BLOOM or CUTOUT for emissive layer). Borderless blocks only have one layer to draw (BLOOM/SOLID). So it's more optimal to treat borderless lamp separately.

**Review by ALongStringOfNumbers — COMMENTED** [review]
Tthe borderless lamps do not have any crafting recipes

**Comment by Tictim** [comment]
> Tthe borderless lamps do not have any crafting recipes

Fixed

**Review by ALongStringOfNumbers — APPROVED** [review]

**Pull request merged**